### PR TITLE
Feature/grasp sensor integration

### PR DESCRIPTION
**Pull request by jlunenburg**
* Added `grasp_sensor_distance` and `object_in_gripper_state` in Arm object
* These are based on Float32 messages received over the `sensor_distance` topic
* The `object_in_gripper_state` property is used by the `Grab` state to detect whether the grasp succeeded
* This can be tested using the `grasp-sensor-sim` script

ToDos
* Test on the real robot (with simulated sensor)
* Test on the real robot with real sensor
* Use distance measurement to improve grasp accuracy: 
    * in `Grab`, reinstate `pre-grasp` (lines 168-176).
    * recompute goal_bl based on the sensor measurement

### CODE REVIEW

**Comment by bobhendrikx** [comment]
@jlunenburg I wont be able to review before end of next week unfortunately (i leave for holiday today). Maybe its better then if someone else does it.

File: robot_skills/src/robot_skills/arms.py [hunk 1]
```diff
@@ -81,6 +90,11 @@ def __init__(self, robot_name, tf_listener, side):
         self._ac_joint_traj = self.create_simple_action_client(
             "/" + robot_name + "/body/joint_trajectory_action", FollowJointTrajectoryAction)
 
+        # Init grasp sensor subscriber
+        self._grasp_sensor_state = (None, rospy.Time.now())
+        rospy.Subscriber("/" + self.robot_name + "/" + self.side + "_gripper/sensor_distance",
```
Comment (bobhendrikx) [inline robot_skills/src/robot_skills/arms.py]: Why not {}_arm/gripper/sensordistance?

File: robot_skills/test/motion/arm-performance [hunk 2]
```diff
@@ -0,0 +1,159 @@
+#! /usr/bin/env python
```
Comment (bobhendrikx) [inline robot_skills/test/motion/arm-performance]: This script is unrelated to the grippersensor, right? maybe put in different branch.

File: robot_skills/src/robot_skills/arms.py [hunk 3]
```diff
@@ -455,6 +507,39 @@ def wait_for_motion_done(self, timeout=10.0, cancel=False):
...
+            return distance
+        else:
+            return None
+
+    def _grasp_sensor_callback(self, msg):
+        """ Callback function for grasp sensor messages
+
+        :param msg: std_msgs.msg.Float32
+        """
+        self._grasp_sensor_state = GripperMeasurement(msg.data[0])
```
Comment (LoyVanBeek) [inline robot_skills/src/robot_skills/arms.py]: What is in the rest of the array ```msg.data```? Wouldn't a plain Float32 make more sense? The messages are a std_msgs.msg.Float32MultiArray according to the constructor

File: robot_skills/src/robot_skills/arms.py [hunk 4]
```diff
@@ -455,6 +469,40 @@ def wait_for_motion_done(self, timeout=10.0, cancel=False):
             # return self._ac_gripper.wait_for_result(rospy.Duration(timeout - passed_time))
             return True
 
+    @property
+    def object_in_gripper_state(self):
```
[the remark this answers is not in the document]
Comment (jlunenburg) [reply robot_skills/src/robot_skills/arms.py]: The states that are using these are responsible for this. This addition is a reflection of the measured gripper state. Should be update the name?

File: robot_skills/src/robot_skills/arms.py [hunk 5]
```diff
@@ -455,6 +469,40 @@ def wait_for_motion_done(self, timeout=10.0, cancel=False):
...
+        """ Returns whether the gripper is empty, holding an object or if this is unknown
```
[the remark this answers is not in the document]
Comment (jlunenburg) [reply robot_skills/src/robot_skills/arms.py]: Fixed

File: robot_skills/src/robot_skills/arms.py [hunk 6]
```diff
@@ -455,6 +469,40 @@ def wait_for_motion_done(self, timeout=10.0, cancel=False):
...
+        distance, stamp = self._grasp_sensor_state
+        if (rospy.Time.now() - stamp).to_sec() < 0.5:
+            return distance
+        else:
+            return None
+
+    def _grasp_sensor_callback(self, msg):
+        """ Callback function for grasp sensor messages
+
+        :param msg: std_msgs.msg.Float32
```
[the remark this answers is not in the document]
Comment (jlunenburg) [reply robot_skills/src/robot_skills/arms.py]: How's that for consistency? We don't do this anywhere... I did update the type: it's a multiarray

File: robot_skills/src/robot_skills/arms.py (hunk 3 above)
Comment (jlunenburg) [reply robot_skills/src/robot_skills/arms.py]: You're right twice. I updated the docstring. I couldn't update the msg type though, since this depends on the available Orocos components. Updating this will take a lot of work and is a completely different issue.

File: test_tools/scripts/grasp-sensor-sim [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+SENSOR_RANGE = (5, 500)  # Range in mm
+
+
+class Side(object):
+    LEFT = "left"
+    RIGHT = "right"
+
+
+class GraspSensorSimulator(QWidget):
+    """ QWidget to simulate the distance sensors on the grippers
```
[the remark this answers is not in the document]
Comment (jlunenburg) [reply test_tools/scripts/grasp-sensor-sim]: Updated

File: test_tools/scripts/grasp-sensor-sim [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+    """
+    def __init__(self):
+        """ Constructor
```
[the remark this answers is not in the document]
Comment (jlunenburg) [reply test_tools/scripts/grasp-sensor-sim]: Updated

File: test_tools/scripts/grasp-sensor-sim [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        # Start a QThread to continuously publish the sensor values
+        self._timer = QTimer(self)
+        self._timer.timeout.connect(self._publish_values)
+        self._timer.start(50)
+
+        # Show the widget
+        self.show()
+
+    def _publish_values(self):
+        """ Publishes the current sensor values (converted from mm to m)
```
[the remark this answers is not in the document]
Comment (jlunenburg) [reply test_tools/scripts/grasp-sensor-sim]: Fixed

File: robot_skills/src/robot_skills/arms.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jlunenburg) [reply robot_skills/src/robot_skills/arms.py]: Renamed to object in gripper measurement in 83184c2e8363a998b91e3b7a6d04d9bec7e4a3d5 . This should be more clear since this reflects a measurement and the state you're talking about is the internal state of this Python session.
